### PR TITLE
Record instructions which are precompiles

### LINKED ISSUE #24705: Inner instruction indexes don't increment for precompiles
#### Problem
In the transaction's metadata we store a record of the inner instructions that were called by a program. Those are matched up to transaction instructions using the index field but transaction instructions that are precompiles don't cause the inner ix index to be incremented.

In the following example, the inner instructions index should be 7, not 6.

https://explorer.solana.com/tx/4oDFDuBTKxth5kyfnpzTNEUnUjYWgEFByYELKb9P1VhEp4QpYE9qWy9Eyn1x5rM5MCDcrKuwrjoJQ7rNQt3fYeK3?cluster=testnet

<details>
  <summary>Example response</summary>

```
{
  "jsonrpc": "2.0",
  "result": {
    "blockTime": 1650901869,
    "meta": {
      "err": null,
      "fee": 10000,
      "innerInstructions": [
        {
          "index": 6,
          "instructions": [
            {
              "parsed": {
                "info": {
                  "destination": "9w61eejrbNTC9CddP5vUVe3QYyQaPKaruP4cguKkmPzZ",
                  "lamports": 5000,
                  "source": "EJUKLLjBMhFnkonfn7wcThnHyDewmhVmG9sEuVP9cvF8"
                },
                "type": "transfer"
              },
              "program": "system",
              "programId": "11111111111111111111111111111111"
            },
            {
              "accounts": [],
              "data": "2kik7JJwphwg1LCJHPyAGRmL9vgKTmkg17wxo45JiAkyn8SruXyzePByitozMpWpeWpJdd4gsBGrf4sxLAXqYhxLsocu2Cjv82mDLaYj21JkkoZysk2pzV1KcYCgNU7",
              "programId": "eeLSJgWzzxrqKv1UxtRVVH8FX3qCQWUs9QuAjJpETGU"
            },
            {
              "accounts": [],
              "data": "LmxsQLBdTJBNF",
              "programId": "eeLSJgWzzxrqKv1UxtRVVH8FX3qCQWUs9QuAjJpETGU"
            }
          ]
        }
      ],
    },
    "slot": 129964372,
    "transaction": {
      "message": {
        "accountKeys": [],
        "addressTableLookups": null,
        "instructions": [
          {
            "accounts": [], // 0
            "data": "16TYTJ8fLSxF",
            "programId": "ComputeBudget111111111111111111111111111111"
          },
          {
            "accounts": [], // 1
            "data": "7YXqSw",
            "programId": "ComputeBudget111111111111111111111111111111"
          },
          {
            "accounts": [], // 2
            "data": "16TYTJ8fLSxF",
            "programId": "ComputeBudget111111111111111111111111111111"
          },
          {
            "accounts": [], // 3
            "data": "7YXqSw",
            "programId": "ComputeBudget111111111111111111111111111111"
          },
          {
            "accounts": [], // 4
            "data": "16TYTJ8fLSxF",
            "programId": "ComputeBudget111111111111111111111111111111"
          },
          {
            "accounts": [], // 5
            "data": "7YXqSw",
            "programId": "ComputeBudget111111111111111111111111111111"
          },
          {
            "accounts": [ // 6
              "KeccakSecp256k11111111111111111111111111111"
            ],
            "data": "2CgVohjrBkqeYWvA",
            "programId": "KeccakSecp256k11111111111111111111111111111"
          },
          {
            "accounts": [ // 7
              "Sysvar1nstructions1111111111111111111111111",
              "EJUKLLjBMhFnkonfn7wcThnHyDewmhVmG9sEuVP9cvF8",
              "9w61eejrbNTC9CddP5vUVe3QYyQaPKaruP4cguKkmPzZ",
              "Hb47E8THZhVs7P4hSWCHsk175qk5vTp2mmqdcRD4pQNb",
              "11111111111111111111111111111111",
              "eeLSJgWzzxrqKv1UxtRVVH8FX3qCQWUs9QuAjJpETGU",
              "6pgKoyyTfRMfrsq8xz3RvmwBUL2GpEe7Z7uBuJE7KYvt",
              "9N3nkyUfmWTvfYvYJi1xVYHW1pyScPsAByccpsoP27S8",
              "3EZXVf4fDhLhrLEyyv6uKkicVnUm5omCzzy5aG6Wnfym",
              "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
            ],
            "data": "3BtaPdCKZxzHJhbrChGuWjc3HGVAzkKZ8QYHPojpKWpd39PCEyPqqEDf29cV1gD2UUiBVRjTnaPXfWLUmr7M3RaPJCW3xVLpUd4bzYMwEwYKUowVXETpWe4SJT7Dy1nzH6Kb5JJ77rydYA8HhEFCs6ckng7Tk5G2SGDB16LammV7LmtEz3BaX9AJDWYhNHHYCj4iaSf33r7QQ29KrLMYwTk754eJ49Ez59wLnKJf",
            "programId": "eeLSJgWzzxrqKv1UxtRVVH8FX3qCQWUs9QuAjJpETGU"
          }
        ],
        "recentBlockhash": "EvRxMd4LeYQxkPQvnTmN7reY3ndQBhtHVgFW5otbAssz"
      },
      "signatures": [
        "4oDFDuBTKxth5kyfnpzTNEUnUjYWgEFByYELKb9P1VhEp4QpYE9qWy9Eyn1x5rM5MCDcrKuwrjoJQ7rNQt3fYeK3"
      ]
    }
  },
  "id": "ef6086fc-0179-48d5-973e-3301119c4491"
}
```

</details>

#### Proposed Solution
Ensure that the inner ix index is incremented for precompiles as well

### PR DESCRIPTION
**Pull request by Lichtso**
#### Problem
See https://github.com/solana-labs/solana/issues/24705

#### Summary of Changes
Don't just skip instructions when `is_precompile` by `continue`,
instead have them be recorded in the `transaction_context`.

Additionally feature gated behind `record_instruction_in_transaction_context_push`.

Fixes #24705
Updates Feature Gate: https://github.com/solana-labs/solana/issues/24965

### CODE REVIEW

**Comment by jstarry** [comment]
Would it be too hacky to do `self.transaction_context.record_precompile_ix()` which directly does `self.instruction_trace.push(vec![])`? Looks like it would work ok here:
```rust
/// Convert from an InstructionTrace to InnerInstructionsList
pub fn inner_instructions_list_from_instruction_trace(
    instruction_trace: &InstructionTrace,
) -> InnerInstructionsList {
    instruction_trace
        .iter()
        .map(|inner_instructions_trace| {
            inner_instructions_trace
                .iter()
                .skip(1)
                .map(|instruction_context| {
```

**Comment by Lichtso** [comment]
Do you mean moving

```Rust
                invoke_context
                    .transaction_context
                    .push(
                        program_indices,
                        &instruction_accounts,
                        &instruction.data,
                        true,
                    )
                    .and_then(|_| invoke_context.transaction_context.pop())
```

into a function `TransactionContext::record_precompile_ix()`?

Or to use [TransactionContext::record_instruction()](https://github.com/solana-labs/solana/blob/71ad12128233045fb782a19846ef927ddc63a2b6/sdk/src/transaction_context.rs#L230) in the `message_processor`?
Because that is already on its way out.

Or, do you want to not actually record anything but sneak in fake entries in `inner_instructions_list_from_instruction_trace()`?
That would cause problems with the sibling instruction introspection I think.

**Comment by jstarry** [comment]
> Or, do you want to not actually record anything but sneak in fake entries in `inner_instructions_list_from_instruction_trace()`?

Yeah, this.

>  That would cause problems with the sibling instruction introspection I think.

Good point, could we protect against this with an enum? Precompiles will always need special handling so might be worth doing something like this:

```rust
pub type InstructionTrace = Vec<InstructionTraceRecord>;

pub enum InstructionTraceRecord {
  Precompile { program_id: Pubkey },
  ProgramInstruction(Vec<InstructionContext>),
}
```

**Comment by Lichtso** [comment]
The enum sounds like a lot more complexity. Keep in mind that in ABIv2 the instruction stack, trace and record are all the same thing. And it is designed to reflect the incoming message as close as possible.

**Comment by jstarry** [comment]
> The enum sounds like a lot more complexity. Keep in mind that in ABIv2 the instruction stack, trace and record are all the same thing. And it is designed to reflect the incoming message as close as possible.

How do precompiles fit into ABIv2?

**Comment by Lichtso** [comment]
That is actually a good question as they were not taken into account yet.
So with the current implementation they would just sit there in the instruction trace (so other instructions can see them) but not be executed. That way the instruction trace stays in sync with the message as long as no error occurs.

File: runtime/src/message_processor.rs [hunk 1]
```diff
@@ -86,10 +89,15 @@ impl MessageProcessor {
             .zip(program_indices.iter())
             .enumerate()
         {
-            if invoke_context
-                .feature_set
-                .is_active(&prevent_calling_precompiles_as_programs::id())
-                && is_precompile(program_id, |id| invoke_context.feature_set.is_active(id))
+            let is_precompile =
+                is_precompile(program_id, |id| invoke_context.feature_set.is_active(id));
```
Comment (jstarry) [inline runtime/src/message_processor.rs:93]: Can you `&&` this with the check that the `prevent_calling_precompiles_as_programs` feature is active? Technically there are no "precompiles" unless that feature is active. We should also ensure that the usage of `is_precompile` below also incorporates the feature gate check.

**Review by jstarry — APPROVED** [review]
Looks good. A bit hard to follow but I think when these feature gates get cleaned up in the future this should get more simplified.
